### PR TITLE
add mae model

### PR DESCRIPTION
**Pull request by Laksh47**
- https://huggingface.co/docs/transformers/v4.34.0/en/custom_models

https://recursion.slack.com/archives/C07S9L58D29/p1729620476934319

### CODE REVIEW

File: mae_utils.py [hunk 1]
```diff
@@ -68,3 +68,23 @@ def unflatten_tokens(
     img = x.reshape(shape=(x.shape[0], -1, h * patch_size, h * patch_size))
 
     return img
+
+def apply_norm_pix(flattened_imgs: torch.Tensor, eps: float = 1.0e-6) -> torch.Tensor:
```
Comment (kian-kd) [inline mae_utils.py]: We can remove this as we do always set this hyperparameter to False!

File: test_huggingface_mae.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+# model = MAEModel.from_pretrained("models/phenom_beta", filename="last.pickle", from_state_dict=True)
+model = MAEModel.from_pretrained("models/phenom_beta", filename="last.pickle")
+
+huggingface_modelpath="recursionpharma/test-pb-model"
+model.push_to_hub(huggingface_modelpath)
+# model.save_pretrained(huggingface_modelpath, push_to_hub=True, repo_id=huggingface_modelpath)
+
+localpath = "models/phenom_beta_huggingface"
+model.save_pretrained(localpath)
+model = MAEModel.from_pretrained(localpath)
```
Comment (kian-kd) [inline test_huggingface_mae.py]: is that localpath the `.safetensors` file?

if so we should add some additional testing logic here to sanity check that the model gives the same outputs when loaded from the `last.pickle` versus when loaded from hugging face.

the inference logic to test is thus:
```
def encoder_test_inference(model):
    with torch.inference_mode():
        x = model.example_input_array.clone()
        x = model.input_norm(x)
        embeddings = model.encoder(x)
    return embeddings
```

then we should do something like this:
```
model_last = MAEModel.from_pretrained("models/phenom_beta", filename="last.pickle")
embeddings_expected = encoder_test_inference(model_last)

model_hugging_face = MAEModel.from_pretrained(localpath)
embeddings_obtained = encoder_test_inference(model_hugging_face)

torch.testing.assert_close(embeddings_expected, embeddings_obtained)
```

File: test_mae.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+from torch.optim.lr_scheduler import OneCycleLR
+from torch.nn import MSELoss, LayerNorm, LazyInstanceNorm2d, Sequential
+from hydra.utils import instantiate
+
+from mae_modules import MAEEncoder, CAMAEDecoder
+from loss import FourierLoss
+from vit import sincos_positional_encoding_vit, vit_small_patch16_256
+from mae import MAE, Normalizer
+
+model_config = {
```
Comment (kian-kd) [inline test_mae.py]: shouldn't this test just load the JSON rather than redefining it here?

File: huggingface_mae.py [hunk 4]
```diff
@@ -0,0 +1,303 @@
+import warnings
```
Comment (kian-kd) [inline huggingface_mae.py]: can we please use normal isort/ruff formatting?

File: models/phenom_beta/config.json [hunk 5]
```diff
@@ -0,0 +1,68 @@
+{
```
Comment (kian-kd) [inline models/phenom_beta/config.json]: why are there two of these config jsons?

File: mae.py [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+def _make_metrics_dict(fourier_loss_weight: float) -> Dict[str, MeanMetric]:
+    metrics = {
+        MAE.TOTAL_LOSS: MeanMetric(),
+        MAE.RECON_LOSS: MeanMetric(),
+    }
+    if fourier_loss_weight > 0:
+        metrics[MAE.FOURIER_LOSS] = MeanMetric()
+    return metrics
+
+class Normalizer(torch.nn.Module):
```
Comment (kian-kd) [inline mae.py]: this is repeated in `normalizer.py`

File: test_mae.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
+import torch
```
Comment (kian-kd) [inline test_mae.py]: it seems like `test_mae` and `test_huggingface_mae` are doing the same thing?

Comment (Laksh47) [reply test_mae.py]: `test_mae` just loads our last.pickle model using torch, `test_huggingface_mae` is where we load it via the huggingface pattern

File: test_mae.py (hunk 3 above)
Comment (Laksh47) [reply test_mae.py]: the way config is defined for our torch model (through hydra) and the hugging face is slightly different. Just keeping it separate

File: huggingface_mae.py (hunk 4 above)
Comment (Laksh47) [reply huggingface_mae.py]: done

File: models/phenom_beta/config.json (hunk 5 above)
Comment (Laksh47) [reply models/phenom_beta/config.json]: one for loading the last.pickle and one for loading the model.safetensors, huggingface expects the config.json to be in the folder

File: mae_utils.py (hunk 1 above)
Comment (Laksh47) [reply mae_utils.py]: done

File: test_huggingface_mae.py (hunk 2 above)
Comment (Laksh47) [reply test_huggingface_mae.py]: done

File: models/phenom_beta_huggingface/config.json [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+  "fourier_loss_weight": 0.0,
+  "input_norm": {
+    "_args_": [
+      {
+        "_target_": "normalizer.Normalizer"
+      },
+      {
+        "_target_": "torch.nn.LazyInstanceNorm2d",
+        "affine": false,
+        "track_running_stats": false
```
Comment (kian-kd) [inline models/phenom_beta_huggingface/config.json]: @Laksh47  can we please change this to the following (not sure if you can do null in JSON - but we need to pass `num_features=None` somehow here:
```suggestion
        "_target_": "torch.nn.InstanceNorm2d",
        "num_features": null,
        "affine": false,
        "track_running_stats": false
```

File: huggingface_mae.py [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+        batch: TensorDict,
+        batch_idx: int,
+        dataloader_idx: int = 0,
+    ) -> None:
+        super().on_validation_batch_end(outputs, batch, batch_idx, dataloader_idx)
+
+    def predict(self, imgs: torch.Tensor) -> torch.Tensor:
+        imgs = self.input_norm(imgs)
+        latent = self.encoder.forward(imgs)
+        return latent
```
Comment (kian-kd) [inline huggingface_mae.py]: CC @okraus this is where we should explore other approaches to improve the Ph-Beta embeddings:
- using `self.encoder_decoder_proj(latent)`
- concatenating token embeddings for the different channels

It seems like doing encoder.forward here will call `self.vit_backbone.forward_head(x)` which I think does the `global_pool` step.

In any case, I would say we can try this:
```
imgs = self.input_norm(imgs)
X = self.encoder.vit_backbone.forward_features(imgs)  # 3d tensor N x num_tokens x dim

# options
if self.apply_decoder_projection_at_inference_time:  # can add as hparam to the model config
    X = self.encoder_decoder_proj(X)

if self.split_tokens_by_channel:  # can add as hparam to the model config
    tokens_per_channel = 256  # phenom-beta has 256 tokens per channel
    N, num_tokens, d = X.shape
    num_channels = num_tokens // tokens_per_channel  
    X_reshaped = X.view(N, num_channels, tokens_per_channel, d)
    pooled_segments = X_reshaped.mean(dim=2)  # Resulting shape: (N, num_channels, d)
    x = pooled_segments.view(N, num_channels * d).contiguous()
else:
    x = self.vit_backbone.pool(X)
return x 
```

**Review by kian-kd — APPROVED** [review]
lets gooo!
